lib/hdb/ext.c

```diff
@@ -530,3 +537,273 @@ hdb_set_last_modified_by(krb5_context context, hdb_entry *entry,
     return 0;
 }
 
+krb5_error_code
+hdb_entry_get_key_rotation(krb5_context context,
+	                   const hdb_entry *entry,
+                           const HDB_Ext_KeyRotation **kr)
+{
+    HDB_extension *ext =
+        hdb_find_extension(entry, choice_HDB_extension_data_key_rotation);
+
+    *kr = ext ? &ext->data.u.key_rotation : NULL;
+    return 0;
+}
+
+krb5_error_code
+hdb_validate_key_rotation(krb5_context context,
+                          const KeyRotation *prev_kr,
+                          const KeyRotation *kr)
+{
+    unsigned int last_kvno = ~0;
+
+    if (kr->period < 1) {
+        krb5_set_error_message(context, EINVAL,
+                               "Key rotation periods must be non-zero "
+                               "and positive");
+        return EINVAL;
+    }
+    if (kr->base_key_kvno < 1 || kr->base_kvno < 1) {
+        krb5_set_error_message(context, EINVAL,
+                               "Key version number zero not allowed "
+                               "for key rotation");
+        return EINVAL;
+    }
+    if (!prev_kr)
+        return 0;
+
+    if (kr->epoch - prev_kr->epoch <= 0) {
```

Comment:
Apparently backwards.  But depends on callers.  Also, we're not yet testing this codepath.